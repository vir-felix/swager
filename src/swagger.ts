import { NextApiRequest, NextApiResponse } from 'next';
import { join } from 'path';
import swaggerJsdoc, { Options } from 'swagger-jsdoc';

type SwaggerOptions = Options & {
  apiFolder?: string;
  schemaFolders?: string[];
  outputFile?: string;
};

const defaultOptions: SwaggerOptions = {
  apiFolder: 'pages/api',
  schemaFolders: [],
  definition: {
    openapi: '3.0.0',
    info: {
      title: 'Next Swagger Doc Demo Api',
      version: '1.0',
    },
  },
};

/**
 * Create swagger JSON
 * @param options.openApiVersion Open API version {3.0.0}
 * @param options.apiFolder NextJS API folder {pages/api}
 * @param options.schemaFolders entity schema folders
 * @param options.title Title
 * @param options.version Version
 * @returns Swagger JSON Spec
 */
export function createSwaggerSpec({
  apiFolder = 'pages/api',
  schemaFolders = [],
  ...swaggerOptions
}: SwaggerOptions = defaultOptions) {
  const scanFolders = [apiFolder, ...schemaFolders];
  const apis = scanFolders.flatMap((folder) => {
    const buildApiDirectory = join(process.cwd(), '.next/server', folder);
    const apiDirectory = join(process.cwd(), folder);
    const publicDirectory = join(process.cwd(), 'public');
    const fileTypes = ['ts', 'tsx', 'jsx', 'js', 'json', 'swagger.yaml'];
    return [
      ...fileTypes.map((fileType) => `${apiDirectory}/**/*.${fileType}`),
      // only scan build directory for *.swagger.yaml and *.js files
      ...['js', 'swagger.yaml', 'json'].map(
        (fileType) => `${buildApiDirectory}/**/*.${fileType}`,
      ),
      // support load static files from public directory
      ...['swagger.yaml', 'json'].map(
        (fileType) => `${publicDirectory}/**/*.${fileType}`,
      ),
    ];
  });
  const options: Options = {
    apis, // files containing annotations as above
    ...swaggerOptions,
  };
  const spec = swaggerJsdoc(options);

  return spec;
}

/**
 * withSwagger middleware
 * @param options.openApiVersion Open API version {3.0.0}
 * @param options.apiFolder NextJS API folder {pages/api}
 * @param options.schemaFolders entity schema folders
 * @param options.title Title
 * @param options.version Version
 * @returns
 */
export function withSwagger({
  apiFolder = 'pages/api',
  schemaFolders = [],
  ...swaggerOptions
}: SwaggerOptions = defaultOptions) {
  return () => (_req: NextApiRequest, res: NextApiResponse) => {
    try {
      const swaggerSpec = createSwaggerSpec({
        apiFolder,
        schemaFolders,
        ...swaggerOptions,
      });
      res.status(200).send(swaggerSpec);
    } catch (error) {
      res.status(400).send(error);
    }
  };
}
