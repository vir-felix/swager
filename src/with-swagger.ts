import { NextApiRequest, NextApiResponse } from 'next';
import { join } from 'path';

import swaggerJsdoc from 'swagger-jsdoc';

type SwaggerOptions = {
  openApiVersion?: string;
  apiFolder?: string;
  title: string;
  version: string;
};

/**
 * Create swagger JSON
 * @param options.openApiVersion Open API version {3.0.0}
 * @param options.apiFolder NextJS API folder {pages/api}
 * @param options.title Title
 * @param options.version Version
 * @returns Swagger JSON Spec
 *
 * @example
 * createSwaggerSpec({
 *  openApiVersion: '3.0.0',
 *  apiFolder:  'pages/api',
 *  title: 'Demo Api',
 *  version: '1.0',
 * })
 */
export function createSwaggerSpec({
  openApiVersion = '3.0.0',
  apiFolder = 'pages/api',
  title,
  version,
}: SwaggerOptions) {
  const apiDirectory = join(process.cwd(), apiFolder);
  const buildApiDirectory = join(process.cwd(), '.next/server', apiFolder);

  const options = {
    definition: {
      openapi: openApiVersion,
      info: {
        title,
        version,
      },
    },
    apis: [
      `${apiDirectory}/**/*.js`,
      `${apiDirectory}/**/*.ts`,
      `${apiDirectory}/**/*.tsx`,
      `${buildApiDirectory}/**/*.js`,
    ], // files containing annotations as above
  };

  return swaggerJsdoc(options);
}

/**
 * withSwagger middleware
 * @param options.openApiVersion Open API version {3.0.0}
 * @param options.apiFolder NextJS API folder {pages/api}
 * @param options.title Title
 * @param options.version Version
 * @returns
 *
 * @example
 * createSwaggerSpec({
 *  openApiVersion: '3.0.0',
 *  apiFolder:  'pages/api',
 *  title: 'Demo Api',
 *  version: '1.0',
 * })
 */
export function withSwagger({
  openApiVersion = '3.0.0',
  apiFolder = 'pages/api',
  title,
  version,
}: SwaggerOptions) {
  return () => {
    return (_req: NextApiRequest, res: NextApiResponse) => {
      try {
        const swaggerSpec = createSwaggerSpec({
          openApiVersion,
          apiFolder,
          title,
          version,
        });
        res.status(200).send(swaggerSpec);
      } catch (error) {
        res.status(400).send(error);
      }
    };
  };
}
