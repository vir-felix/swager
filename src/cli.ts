import { cli } from 'cleye';
import { readFileSync, writeFileSync } from 'fs';

import { createSwaggerSpec } from './swagger';

// Parse argv
const argv = cli({
  name: 'next-swagger-doc-cli',

  // Define parameters
  // Becomes available in ._.filePath
  parameters: [
    '<config file>', // next swagger config file is required
  ],

  // Define flags/options
  // Becomes available in .flags
  flags: {
    // Parses `--output` as a string
    output: {
      type: String,
      description: 'Output file path',
      default: 'public/swagger.json',
    },
  },
});

const config = readFileSync(argv._.configFile);

const spec: Record<string, any> = createSwaggerSpec(
  JSON.parse(config.toString()),
);

console.log(
  `Generating swagger spec to ${argv.flags.output} with config`,
  config.toString(),
);
writeFileSync(argv.flags.output, JSON.stringify(spec, null, 2));
