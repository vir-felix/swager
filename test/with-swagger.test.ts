import { createSwaggerSpec } from '../src/with-swagger';

describe('withSwagger', () => {
  it('should create default swagger json option', () => {
    expect(
      createSwaggerSpec({
        title: 'NextJS Swagger',
        version: '0.1.0',
      })
    ).toMatchSnapshot();
  });
});
