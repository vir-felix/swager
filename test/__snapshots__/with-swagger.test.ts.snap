// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSwagger should create default swagger json option 1`] = `
Object {
  "components": Object {},
  "info": Object {
    "title": "NextJS Swagger",
    "version": "0.1.0",
  },
  "openapi": "3.0.0",
  "paths": Object {},
  "tags": Array [],
}
`;
